server/src/main/java/org/elasticsearch/action/DocWriteRequest.java

```diff
@@ -195,16 +197,22 @@ public static OpType fromString(String sOpType) {
         }
     }
 
-    /** read a document write (index/delete/update) request */
-    static DocWriteRequest<?> readDocumentRequest(StreamInput in) throws IOException {
+    /**
+     * Read a document write (index/delete/update) request
+     *
+     * @param shardId shard id of the request. {@code null} when reading as part of a {@link org.elasticsearch.action.bulk.BulkRequest}
```

Comment:
Having to maintain the old format for 8.x (here and all throughout this change set pretty much) is quite annoying but I think unavoidable because we need to continue to support a `7.x` `TransportClient` sending us a `BulkRequest`.

Reply:
That sounds suspicious. The transport client is deprecated in 7.x, do we expect it to work at all against a cluster containing v8 nodes? Even if we do, do we expect it to work even if the client version is earlier than 7.last? Perhaps I'm missing something here...

Reply:
>  The transport client is deprecated in 7.x, do we expect it to work at all against a cluster containing v8 nodes?

Good question. I guess as of today it will be able to connect to a v8 cluster at least with 7.last.

> Even if we do, do we expect it to work even if the client version is earlier than 7.last?

No I don't think so. Only 7.last should work, ES won't connect to anything older.

=> still ... it would be my understanding that we have to support this request in 8.x unless we patch the transport client in 7.last to not connect to 8.x nodes. (I'd be all for that)

Still no matter what we do there, I'd keep this compatibility step here for this PR so I can backport it to 7.x as is. We can then look into if and how we want to go about removing over the wire serialization for `BulkRequest` in 8.x in a separate issue?